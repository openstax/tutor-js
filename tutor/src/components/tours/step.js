import {
    React, PropTypes, observer,
} from 'vendor';
import Spotlight from './spotlight';
import Floater from 'react-floater';
import STEPS from './custom';
import { colors } from 'theme';

@observer
export default class TourStep extends React.Component {

    static propTypes = {
        step: PropTypes.object.isRequired,
        ride: PropTypes.object.isRequired,
    }

    styles = {
        floater: {
            maxWidth: '100%',
        },
        container: {
            padding: 0,
            borderRadius: '4px',
        },
        options: {
            zIndex: 1501,
        },
    };

    classNameStyles = {
        green: {
            arrow: {
                color: colors.secondary,
            },
        },
        greenNoBody: {
            arrow: {
                color: colors.secondary,
                margin: 14,
            },
            container: {
                minHeight: 0,
                padding: 0,
            },
        },
    };


    render() {
        const { step, ride } = this.props;
        const Step = STEPS[step.customComponent || 'Standard'];
        const styles = { ...this.styles, ...this.classNameStyles[step.className] };

        const tip = (
            <Floater
                open={true}
                debug={ride.context.emitDebugInfo}
                autoOpen={true}
                key={step.anchor_id}
                target={step.target}
                styles={styles}
                disableAnimation={true}
                placement={step.placement}
                content={<Step step={step} ride={ride} />}
            />
        );
        if (step.shouldShowSpotlight) {
            return (
                <Spotlight step={step} ride={ride}>
                    {tip}
                </Spotlight>
            );
        }
        return tip;
    }

}
