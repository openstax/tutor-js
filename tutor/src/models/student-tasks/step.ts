import {
    BaseModel, field, observable, computed, array, action, model, modelize,
    NEW_ID, lazyGetter, hydrateModel, getParentOf,
} from 'shared/model';
import S from '../../helpers/string';
import { pick, get, isNil, omit } from 'lodash';
import { extractCnxId } from '../../helpers/content';
import { MediaActions } from '../../flux/media';
import Time from 'shared/model/time';
import urlFor from '../../api';
import {
    ReferenceBookNode, RelatedContent, ChapterSection, Exercise, StudentTask, StudentTaskStepGroup,
} from '../../models';

class TaskStepContent extends BaseModel {
    update(data: any) {
        Object.assign(this, data);
    }
    requiresAnswerId = false;
}

class StudentTaskVideoStep extends TaskStepContent { }
class StudentTaskExternalStep extends TaskStepContent { }
class StudentTaskInteractiveStep extends TaskStepContent { }
class StudentTaskPlaceHolderStep extends TaskStepContent { }

class StudentTaskReadingStep extends TaskStepContent {
    title = '';
    content_url = ''
    html = ''
    chapter_section = ''

    constructor() {
        super()
        modelize(this)
    }

    @model(RelatedContent) related_content = array<RelatedContent>()

    @lazyGetter get chapterSection() { return new ChapterSection(this.chapter_section) }
    @lazyGetter get page() {
        return hydrateModel(ReferenceBookNode, {
            uuid: this.related_content[0].uuid,
            id: this.related_content[0].page_id,
            cnx_id: extractCnxId(this.content_url),
            content_html: this.html,
            title: this.related_content[0].title,
            chapter_section: this.chapterSection,
        })
    }

    @computed get pageTitle(): string {
        return this.title || get(this, 'related_content[0].title');
    }
}

export
class StudentTaskExerciseStep extends Exercise {
    constructor() {
        super()
        modelize(this)
    }
    get stem_html() { return this.content.stem_html; }
    get questions() { return this.content.questions; }
    get stimulus_html() { return this.content.stimulus_html; }
    get requiresAnswerId() { return this.content.isMultiChoice; }
}

const ContentClasses = {
    video: StudentTaskVideoStep,
    reading: StudentTaskReadingStep,
    exercise: StudentTaskExerciseStep,
    external_url: StudentTaskExternalStep,
    placeholder: StudentTaskPlaceHolderStep,
    interactive: StudentTaskInteractiveStep,
};

const UNSAVEABLE_TYPES = [
    'placeholder',
];

const NO_ADDITIONAL_CONTENT = [
    'external_url',
];

export class StudentTaskStep extends BaseModel {

    @field id = NEW_ID;
    @field uid = '';
    @field preview = '';
    @field available_points = 0;
    @field type = '';
    @field is_completed = false
    @field answer_id? = NEW_ID;
    @field free_response = '';
    @field feedback_html = '';
    @field correct_answer_id = NEW_ID;
    @model(Time) last_completed_at = Time.unknown
    @field response_validation?: any = {}
    @field spy?: any = {}
    @field external_url = ''
    @field labels?: any[];
    @field formats?: any[];
    @field group = ''
    @field can_be_updated = false
    @field dropped_method = ''
    @field is_feedback_available = false
    @field published_points = 0
    @field published_comments = ''
    @field published_points_without_lateness = 0
    @field published_late_work_point_penalty = 0
    @field tasked_id = NEW_ID
    @field exercise_id = NEW_ID

    @observable content?: any
    @observable isFetched = false

    @observable multiPartGroup?: StudentTaskStepGroup

    constructor() {
        super()
        modelize(this)
    }

    get task() { return getParentOf<StudentTask>(this) }

    @computed get canAnnotate() {
        return this.isReading;
    }

    @computed get chapterSection() {
        return this.content.related_content ?
            this.content.related_content[0].chapter_section : null;
    }

    @computed get isExercise() {
        return 'exercise' === this.type;
    }
    @computed get isReading() {
        return 'reading' === this.type;
    }
    @computed get isExternalUrl() {
        return 'external_url' === this.type;
    }
    @computed get isInteractive() {
        return 'interactive' === this.type;
    }
    @computed get isVideo() {
        return 'video' === this.type;
    }
    @computed get isPlaceHolder() {
        return 'placeholder' === this.type;
    }
    @computed get isDroppedQuestion() {
        return Boolean(this.dropped_method);
    }

    // read properties from content when it may not have been fetched yet
    readContentProperty(property: string) {
        if (!this.content) {
            return null;
        }
        return this.content[property];
    }

    @computed get isOpenEndedExercise() {
        return Boolean(this.isExercise && this.readContentProperty('isOpenEnded'));
    }

    @computed get isWrittenResponseExercise() {
        return Boolean(this.isExercise && this.readContentProperty('isWrittenResponse'));
    }

    @computed get isCorrect() {
        return Boolean(
            this.correct_answer_id && this.answer_id == this.correct_answer_id
        );
    }

    @computed get pointsScored() {
        if(!isNil(this.published_points)) return this.published_points;
        if (this.correct_answer_id) {
            return this.answer_id === this.correct_answer_id
                ? this.available_points
                : 0;
        }
        return null;
    }

    @computed get isTwoStep() {
        return Boolean(
            this.isExercise && this.readContentProperty('isMultiChoice') && this.formats?.includes('free-response'),
        )
    }

    @computed get isReview() {
        return this.labels?.includes('review');
    }

    @computed get isPersonalized() {
        return 'personalized' == this.group ;
    }

    @computed get isSpacedPractice() {
        return 'spaced practice' == this.group ;
    }

    @computed get canEditFreeResponse() {
        return Boolean(
            this.can_be_updated && !this.answer_id &&
                this.formats?.includes('free-response') &&
                (this.readContentProperty('isOpenEnded') || S.isEmpty(this.free_response))
        );
    }

    @computed get hasBeenAnswered() {
        return Boolean(this.answer_id);
    }

    @computed get canAnswer() {
        return Boolean(
            this.isExercise && this.can_be_updated
        );
    }

    @computed get needsFetched() {
        return Boolean(
            !NO_ADDITIONAL_CONTENT.includes(this.type) && !this.api.hasBeenFetched
        );
    }

    @computed get isLate() {
        return this.last_completed_at.isAfter(this.task.due_at)
    }

    @action fetchIfNeeded() {
        if (this.needsFetched && !this.api.isInProgress('fetchStudentTaskStep')) {
            this.fetch();
        }
    }

    @action markViewed() {
        if (this.can_be_updated && !this.isExercise && !this.isExternalUrl && !this.is_completed) {
            this.is_completed = true;
            this.save();
        }
    }

    // called when the task is reloaded and each step is reset
    @action setFromTaskFetch(data: any) {
        // the step is being re-used and it's type changed
        if (data.id != this.id || data.type != this.type) {
            this.content = null
            this.isFetched = false
        }
        this.api.reset();
        this.update(data);
    }

    // called by API
    async fetch(): Promise<void> {
        const data = await this.api.request(urlFor('fetchStudentTaskStep', { stepId: this.id }))
        this.onLoaded(data)
    }

    async save(): Promise<void> {
        if (UNSAVEABLE_TYPES.includes(this.type)) { return; }
        const data = await this.api.request(
            urlFor('saveStudentTaskStep', { stepId: this.id }),
            { data: pick(this, 'is_completed', 'answer_id', 'free_response', 'response_validation') }
        )
        this.onLoaded(data)
    }

    @action beginRecordingAnswer({ free_response }: { free_response: string }) {
        this.free_response = free_response;
        if (this.content.requiresAnswerId) {
            this.answer_id = undefined;
        } else {
            this.is_completed = true;
        }
    }

    @action onLoaded(data: any) {
        this.update(omit(data, 'id'))
        const Klass = ContentClasses[this.type];
        if (!Klass) {
            throw new Error(`Attempted to set content on unknown step type ${this.type}`);
        }
        this.content = hydrateModel(Klass, data, this)
        if (this.isReading) {
            MediaActions.parse(this.content.html);
        }
        this.isFetched = true;
    }

}
