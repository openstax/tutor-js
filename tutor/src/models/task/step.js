import {
  BaseModel, identifiedBy, identifier, hasMany, field, belongsTo,
} from '../base';
import { computed } from 'mobx';
import { last } from 'lodash';

import RelatedContent from './step-related-content';

@identifiedBy('task/step')
export default class TaskStep extends BaseModel {

  @identifier id;
  @field task_id;
  @field title;
  @field type;

  @field({ type: 'array' }) chapter_section;
  @field content_html;
  @field content_url;
  @field({ type: 'date' }) first_completed_at;
  @field({ type: 'date' }) last_completed_at;

  @field group;
  @field has_recovery;

  @field is_completed;

  @field({ type: 'array' }) labels;

  @hasMany({ model: RelatedContent }) related_content;

  @belongsTo({ model: 'task' }) task;

  @computed get cnxId() {
    return last( (this.content_url || '').split('contents/') );
  }
}
