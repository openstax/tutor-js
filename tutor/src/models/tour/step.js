import {
  BaseModel, identifiedBy, identifier, belongsTo, field,
} from 'shared/model';

import {
  computed, action,
} from 'mobx';

import { isEmpty, intersection } from 'lodash';

import Markdown from 'markdown-it';
import MDRegex from 'markdown-it-regexp';

import Actions from './actions';
import WindowSize from '../window-size';

const MD = Markdown({ html: true, linkify: true, typographer: true });

MD.use(MDRegex(/:best-practices:/, () => '<i class="tour-step-best-practices"></i>' ));

// TourStep
// A step in a tour, where steps are connected by a “next” button that leads from one step to the next.
// Can be linked to either an anchor or region
// Has a title and rich text body.

@identifiedBy('tour/step')
export default class TourStep extends BaseModel {
  @identifier id;

  @belongsTo tour;

  @field title;
  @field body;
  @field position;
  @field isCancelable = true;
  @field is_fixed;
  @field anchor_id;
  @field customComponent;
  @field spotlight = true;
  @field displayAs = 'standard';
  @field spotLightPadding = 5;
  @field requiredViewsCount = 1;
  @field displayWithButtons = true;
  @field({ type: 'object' }) action;
  @field className;
  @field({ type: 'array' }) disabledBreakpoints;

  windowSize = new WindowSize();

  @computed get target() {
    return this.anchor_id ? `[data-tour-anchor-id="${this.anchor_id}"]` : null;
  }

  get element() {
    return this.target ? document.querySelector(this.target) : null;
  }

  @computed get placement() {
    if (!this.element) { return 'center'; }
    return (this.anchor_id && this.position) ? this.position : 'auto';
  }

  @computed get isCentered() {
    return this.position == 'center';
  }

  @computed get actionClass() {
    if (!this.action) { return null; }
    return Actions.forIdentifier(this.action.id);
  }

  @computed get actionInstance() {
    return this.actionClass && new this.actionClass({
      step: this,
      ...this.action,
    });
  }

  @computed get shouldShowSpotlight() {
    return Boolean((this.anchor_id || this.isCentered) && this.spotlight);
  }

  @action preValidate() {
    this.actionInstance && this.actionInstance.preValidate();
  }

  @action prepare(options) {
    return (this.actionInstance && this.actionInstance.beforeStep(options)) || Promise.resolve();
  }

  @action complete(options) {
    return (this.actionInstance && this.actionInstance.afterStep(options) ) || Promise.resolve();
  }

  get isViewable() {
    if (!isEmpty(intersection(this.disabledBreakpoints, [this.windowSize.currentBreakpoint]))) {
      return false;
    }
    return Boolean(!this.target || this.element);
  }

  @computed get HTML() {
    return this.body ? MD.render(this.body) : '';
  }

  @computed get shouldReplay() {
    return this.requiredViewsCount > this.tour.viewCounts;
  }

}
