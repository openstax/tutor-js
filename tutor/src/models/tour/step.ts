import { BaseModel, field, modelize, computed, action, NEW_ID, getParentOf } from 'shared/model';
import { isEmpty, intersection } from 'lodash';
import Markdown from 'markdown-it';
import MDRegex from 'markdown-it-regexp';
import Actions from './actions';
import WindowSize from '../window-size';
import type Tour from '../tour'

const MD = Markdown({ html: true, linkify: true, typographer: true });
MD.use(MDRegex(/:best-practices:/, () => '<i class="tour-step-best-practices"></i>' ))

// TourStep
// A step in a tour, where steps are connected by a “next” button that leads from one step to the next.
// Can be linked to either an anchor or region
// Has a title and rich text body.

export default class TourStep extends BaseModel {
    @field id = NEW_ID;

    @field title = '';
    @field body = '';
    @field position = '';
    @field isCancelable = true;
    @field is_fixed = false;
    @field anchor_id: string | null = null;
    @field customComponent: any;
    @field spotlight = true;
    @field displayAs = 'standard';
    @field spotLightPadding = 5;
    @field requiredViewsCount = 1;
    @field displayWithButtons = true;
    @field action?: any;
    @field className?: string;
    @field disabledBreakpoints?: any[];

    windowSize = new WindowSize();

    constructor() {
        super();
        modelize(this);
    }

    get tour() { return getParentOf<Tour>(this) }

    @computed get target() {
        return this.anchor_id ? `[data-tour-anchor-id="${this.anchor_id}"]` : null;
    }

    get element(): HTMLElement | null {
        return this.target ? document.querySelector(this.target) : null;
    }

    @computed get placement() {
        if (!this.element) { return 'center'; }
        return (this.anchor_id && this.position) ? this.position : 'auto';
    }

    @computed get isCentered() {
        return this.position == 'center';
    }

    @computed get actionClass() {
        if (!this.action) { return null; }
        return Actions.forIdentifier(this.action.id);
    }

    @computed get actionInstance() {
        return this.actionClass && new this.actionClass({
            step: this,
            ...this.action,
        });
    }

    @computed get shouldShowSpotlight() {
        return Boolean((this.anchor_id || this.isCentered) && this.spotlight);
    }

    @action preValidate() {
        this.actionInstance && this.actionInstance.preValidate();
    }

    @action prepare(options: any) {
        return (this.actionInstance && this.actionInstance.beforeStep(options)) || Promise.resolve();
    }

    @action complete(options: any) {
        return (this.actionInstance && this.actionInstance.afterStep(options) ) || Promise.resolve();
    }

    get isViewable() {
        if (!isEmpty(intersection(this.disabledBreakpoints, [this.windowSize.currentBreakpoint]))) {
            return false;
        }
        return Boolean(!this.target || this.element);
    }

    @computed get HTML() {
        return this.body ? MD.render(this.body) : '';
    }

    @computed get shouldReplay() {
        return this.requiredViewsCount > this.tour.viewCounts;
    }
}
