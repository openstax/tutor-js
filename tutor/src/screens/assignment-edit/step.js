import { React, PropTypes, observer, observable, computed, action } from 'vendor';
import details from './details';
import chapters from './chapters';
import questions from './questions';
import points from './points';
import reorder from './reorder';

import UnknownType from './unknown';

const reading = {
  details,
  chapters,
  reorder,
};

const homework = {
  details,
  chapters,
  questions,
  points,
};

const external = {
  details,
};

const event = {
  details,
};

const StepTypes = {
  homework, reading, external, event,
};


const Step = observer(({ ux }) => {
  const Component = ux.steps.currentStepComponent || UnknownType;
  return <Component ux={ux} />;
});

Step.propTypes = {
  ux: PropTypes.object.isRequired,
};

class StepUX {

  constructor(ux) {
    this.ux = ux;
  }

  @observable _stepIndex = 0;

  @computed get number() {
    return this._stepIndex + 1;
  }

  @computed get headerText() {
    return this.hasSteps ? `STEP ${this.number}` : null;
  }

  @computed get hasSteps() {
    return this.stepIds.length > 1;
  }

  @action.bound async goForward() {
    if (this.canGoForward) {
      // TODO, skip steps if the assignment type doesn't need the next,
      // for instance events won't have chapters & questiosn
      await this.ux.saveAsDraft();

      this.ux.navigateToStep(this.stepIds[this._stepIndex + 1]);
    }
  }

  @action setIndex(index) {
    const stepIndex = this.stepIds.indexOf(index);
    if (-1 !== stepIndex) {
      this._stepIndex = stepIndex;
    }
  }

  @action.bound goBackward() {
    if (this.canGoBackward) {
      this.ux.navigateToStep(this.stepIds[this._stepIndex - 1]);
    }
  }

  @computed get isLast() {
    return this._stepIndex >= this.stepIds.length - 1;
  }

  @computed get isFirst() {
    return this._stepIndex <= 0;
  }

  @computed get stepComponents() {
    return StepTypes[this.ux.plan.type];
  }

  @computed get stepIds() {
    return Object.keys(this.stepComponents || {});
  }

  @computed get currentStepComponent() {
    return this.stepComponents[this.currentStepId];
  }

  @computed get currentStepId() {
    return this.stepIds[this._stepIndex];
  }

  @computed get canGoForward() {
    return Boolean(
      !this.isApiPending &&
        !this.isLast &&
        this.ux.validations.isValid
    );
  }

  @computed get canGoBackward() {
    return Boolean(
      !this.isApiPending && !this.isFirst
    );
  }

  @computed get canSubmit() {
    return Boolean(
      this.ux.validations.isValid
    );
  }

}


export { StepUX, Step, StepTypes };
