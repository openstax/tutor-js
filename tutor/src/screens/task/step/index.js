import {
  React, PropTypes, observer, styled,
} from 'vendor';
import {
  Facebook as ReadingLoader,
  BulletList as HomeworkLoader,
} from 'react-content-loader';
import Reading from './reading';
import Exercise from './exercise';
import Placeholder from './placeholder';
import HtmlContent from './html-content';
import Failure from './failure';
import End from './end';
import { LoadingCard } from './card';
import {
  PersonalizedGroup,
  TwoStepValueProp,
  IndividualReview,
  SpacedPractice,
} from './value-props';


const STEP_TYPES = {
  end: End,
  reading: Reading,
  video: HtmlContent,
  exercise: Exercise,
  placeholder: Placeholder,
  interactive: HtmlContent,
  'two-step-intro': TwoStepValueProp,
  'personalized-intro': PersonalizedGroup,
  'spaced-practice-intro': SpacedPractice,
  'individual-review-intro': IndividualReview,
};

const PENDING_TYPES = {
  reading: ReadingLoader,
  exercise: HomeworkLoader,
  video: HtmlContent.Loader,
  interactive: HtmlContent.Loader,
};

// has plenty of space at bottom so last step can
// be scrolled up and user knows it's last
const MultipartGroup = styled.div`
  padding-bottom: 200px;
`;

@observer
class TaskStep extends React.Component {

  static propTypes = {
    pending: PropTypes.func,
    ux: PropTypes.object.isRequired,
    step: PropTypes.shape({
      steps: PropTypes.array,
      type: PropTypes.string.isRequired,
      needsFetched: PropTypes.bool,
      api: PropTypes.shape({
        hasErrors: PropTypes.bool,
      }),
    }).isRequired,
  };

  render() {
    const { ux, step } = this.props;

    if (!step || (step.api && step.api.hasErrors)) {
      return <Failure task={ux.task} step={step} />;
    }
    const { type, needsFetched } = step;

    const stepProps = {
      ...this.props,
      onContinue: ux.canGoForward ? ux.goForward : null,
    };

    if ('mpq' === type) {
      return (
        <MultipartGroup>
          {step.steps.map((s, i) =>
            <TaskStep
              key={i}
              {...stepProps}
              isMultiPart
              isFollowupMPQ={0 !== i}
              step={s}
            />)}
        </MultipartGroup>
      );
    }

    if (needsFetched) {
      const Pending = PENDING_TYPES[type] || PENDING_TYPES.reading;
      return <LoadingCard><Pending /></LoadingCard>;
    }

    const Step = STEP_TYPES[type];

    if (!Step) {
      return <Failure task={ux.task} step={step} />;
    }

    return (
      <Step {...stepProps} />
    );
  }
}

export { TaskStep };
