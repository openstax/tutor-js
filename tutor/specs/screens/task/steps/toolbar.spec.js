import { StepToolbar } from '../../../../src/screens/task/step/toolbar';
import { Factory, TimeMock, C } from '../../../helpers';

describe('Task Step Toolbar', () => {
    let props;
    TimeMock.setTo('2017-10-14T12:00:00.000Z');

    beforeEach(() => {
        const task = Factory.studentTask({ type: 'homework', stepCount: 1 });
        props = {
            course: Factory.course(), step: task.steps[0],
        };
    });

    it('matches snapshot', () => {
        const bar = mount(<C><StepToolbar {...props} /></C>);
        expect(bar.debug()).toMatchSnapshot();
        bar.unmount();
    });

    it('hides the toolbar', () => {
      const bar = mount(<C><StepToolbar hideToolbar={true} {...props} /></C>);
      expect(bar).toEqual({});
      bar.unmount();
    });

    it('hides save to practice button', () => {
      const bar = mount(<C><StepToolbar showSaveToPractice={false} {...props} /></C>);
      expect(bar).not.toHaveRendered('SavePracticeButton');
      bar.unmount();
    });

    it('hides content link', () => {
      const bar = mount(<C><StepToolbar hideContentLink={true} {...props} /></C>);
      expect(bar).not.toHaveRendered('BrowseTheBook');
      bar.unmount();
    });
});
