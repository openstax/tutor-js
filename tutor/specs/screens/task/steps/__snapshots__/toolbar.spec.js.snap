// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Step Toolbar matches snapshot 1`] = `
"<C>
  <Component theme={{...}}>
    <ForwardRef noRef={[undefined]} path=\\"/\\" withRef={false}>
      <MemoryRouter initialIndex={0} initialEntries={{...}}>
        <Router history={{...}}>
          <TestContextWrapper>
            <Memo() backend={[Function: createBackend]}>
              <Component>
                <SpyModeWrapper>
                  <MobXProvider spyMode={{...}}>
                    <div className=\\"openstax-debug-content\\">
                      <ModalManager>
                        <MobXProvider modalManager={{...}}>
                          <inject-with-modalManager-spyMode(TourConductor) tourContext={[undefined]}>
                            <TourConductor tourContext={[undefined]} modalManager={{...}} spyMode={{...}}>
                              <MobXProvider tourContext={{...}}>
                                <div data-purpose=\\"tour-conductor-wrapper\\">
                                  <TutorSpecLayout>
                                    <MobXProvider topNavbar={{...}} courseContext={{...}} bottomNavbar={{...}} setSecondaryTopControls={[Function: setSecondaryTopControls] { isMobxAction: true }}>
                                      <div>
                                        <Navbar area=\\"header\\" context={{...}} isDocked={false} />
                                        <StepToolbar course={{...}} step={{...}}>
                                          <styled.div>
                                            <div className=\\"sc-dkPtyc fPgPLT\\">
                                              <BrowseTheBook unstyled={true} course={{...}} chapterSection={{...}} tag=\\"a\\" windowImpl={{...}}>
                                                <a data-test-id=\\"browse-the-book-btn\\" aria-label=\\"Browse course reference book\\" onClick={[Function: onClick] { isMobxAction: true }} href=\\"/book/2/section/1.2\\" target=\\"_blank\\" className=\\"browse-the-book\\">
                                                  <Icon type=\\"book-open\\" buttonProps={{...}} tooltipProps={{...}}>
                                                    <IconWrapper data-variant={[undefined]} color={[undefined]} hoverColor={[undefined]} className=\\"ox-icon ox-icon-book-open\\" icon={{...}}>
                                                      <FontAwesomeIcon data-variant={[undefined]} color={[undefined]} className=\\"IconWrapper-ra27cp-0 dbDpez ox-icon ox-icon-book-open\\" icon={{...}} border={false} mask={{...}} fixedWidth={false} inverse={false} flip={{...}} listItem={false} pull={{...}} pulse={false} rotation={{...}} size={{...}} spin={false} symbol={false} title=\\"\\" transform={{...}} swapOpacity={false}>
                                                        <svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"book-open\\" className=\\"svg-inline--fa fa-book-open fa-w-18 IconWrapper-ra27cp-0 dbDpez ox-icon ox-icon-book-open\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 576 512\\" style={{...}} data-variant={[undefined]} color={[undefined]}>
                                                          <path fill=\\"currentColor\\" d=\\"M542.22 32.05c-54.8 3.11-163.72 14.43-230.96 55.59-4.64 2.84-7.27 7.89-7.27 13.17v363.87c0 11.55 12.63 18.85 23.28 13.49 69.18-34.82 169.23-44.32 218.7-46.92 16.89-.89 30.02-14.43 30.02-30.66V62.75c.01-17.71-15.35-31.74-33.77-30.7zM264.73 87.64C197.5 46.48 88.58 35.17 33.78 32.05 15.36 31.01 0 45.04 0 62.75V400.6c0 16.24 13.13 29.78 30.02 30.66 49.49 2.6 149.59 12.11 218.77 46.95 10.62 5.35 23.21-1.94 23.21-13.46V100.63c0-5.29-2.62-10.14-7.27-12.99z\\" style={{...}} />
                                                        </svg>
                                                      </FontAwesomeIcon>
                                                    </IconWrapper>
                                                  </Icon>
                                                  <span>
                                                    View topic in textbook
                                                  </span>
                                                </a>
                                              </BrowseTheBook>
                                              <ExerciseSuggestCorrectionLink exerciseId=\\"1@1\\" bookUUID=\\"8d04a686-d5e8-4798-a27d-c608e4d0e187\\" related_content={{...}}>
                                                <a target=\\"_blank\\" tabIndex={-1} href=\\"https://openstax.org/errata/form?source=tutor&location=1%401%201.2%20Newton%27s%20First%20Law%20of%20Motion%3A%20Inertia&book=The%20AP%20Physics%20Collection\\">
                                                  <Icon type=\\"exclamation-circle\\" buttonProps={{...}} tooltipProps={{...}}>
                                                    <IconWrapper data-variant={[undefined]} color={[undefined]} hoverColor={[undefined]} className=\\"ox-icon ox-icon-exclamation-circle\\" icon={{...}}>
                                                      <FontAwesomeIcon data-variant={[undefined]} color={[undefined]} className=\\"IconWrapper-ra27cp-0 dbDpez ox-icon ox-icon-exclamation-circle\\" icon={{...}} border={false} mask={{...}} fixedWidth={false} inverse={false} flip={{...}} listItem={false} pull={{...}} pulse={false} rotation={{...}} size={{...}} spin={false} symbol={false} title=\\"\\" transform={{...}} swapOpacity={false}>
                                                        <svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"exclamation-circle\\" className=\\"svg-inline--fa fa-exclamation-circle fa-w-16 IconWrapper-ra27cp-0 dbDpez ox-icon ox-icon-exclamation-circle\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\" style={{...}} data-variant={[undefined]} color={[undefined]}>
                                                          <path fill=\\"currentColor\\" d=\\"M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z\\" style={{...}} />
                                                        </svg>
                                                      </FontAwesomeIcon>
                                                    </IconWrapper>
                                                  </Icon>
                                                  <span>
                                                    Suggest a correction
                                                  </span>
                                                </a>
                                              </ExerciseSuggestCorrectionLink>
                                            </div>
                                          </styled.div>
                                        </StepToolbar>
                                        <Navbar area=\\"footer\\" context={{...}} />
                                      </div>
                                    </MobXProvider>
                                  </TutorSpecLayout>
                                  <SpyModeContent>
                                    <div className=\\"visible-when-debugging\\">
                                      <div className=\\"tour-spy-info\\">
                                        available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
                                      </div>
                                    </div>
                                  </SpyModeContent>
                                </div>
                              </MobXProvider>
                            </TourConductor>
                          </inject-with-modalManager-spyMode(TourConductor)>
                        </MobXProvider>
                      </ModalManager>
                      <a href=\\"#spy\\" tabIndex={-1} onClick={[Function: toggleDebug] { isMobxAction: true }} className=\\"debug-toggle-link\\" aria-hidden=\\"true\\">
                        Ï€
                      </a>
                    </div>
                  </MobXProvider>
                </SpyModeWrapper>
              </Component>
            </Memo()>
          </TestContextWrapper>
        </Router>
      </MemoryRouter>
    </ForwardRef>
  </Component>
</C>"
`;
